pkg/eigenState/stakerShares/stakerShares.go

```diff
@@ -610,88 +653,172 @@ func (ss *StakerSharesModel) HandleStateChange(log *storage.TransactionLog) (int
-	`
-	existingRecords := make([]StakerShares, 0)
-	res := ss.DB.Model(&StakerShares{}).
-		Raw(query,
-			sql.Named("slotIds", slotIds),
-		).
-		Scan(&existingRecords)
+			// append the delta to the delta accumulator
+			ss.deltaAccumulator[blockNumber] = append(ss.deltaAccumulator[blockNumber], shareDelta)
 
-	if res.Error != nil {
-		ss.logger.Sugar().Errorw("Failed to fetch staker_shares", zap.Error(res.Error))
-		return nil, res.Error
-	}
+			currStakerShares, ok := updatedStakerShares[slotID]
+			if !ok {
+				stakerShares := &StakerShares{}
+				// get the record from the database with the given staker and strategy with the latest blockNumber
+				res := ss.DB.Model(&StakerShares{}).Where("staker = ? AND strategy = ?", shareDelta.Staker, shareDelta.Strategy).Order("block_number desc").First(&stakerShares)
+				if res.Error != nil && res.Error != gorm.ErrRecordNotFound {
+					ss.logger.Sugar().Errorw("Failed to fetch staker_shares", zap.Error(res.Error))
+					return nil, res.Error
+				}
 
-	// Map the existing records to a map for easier lookup
-	mappedRecords := make(map[types.SlotID]StakerShares)
-	for _, record := range existingRecords {
-		slotId := NewSlotID(record.Staker, record.Strategy)
-		mappedRecords[slotId] = record
-	}
+				// if the record is not found, create a new record with shares as 0
+				if res.Error == gorm.ErrRecordNotFound {
+					stakerShares = &StakerShares{
+						Staker:      shareDelta.Staker,
+						Strategy:    shareDelta.Strategy,
+						Shares:      "0",
+						BlockNumber: blockNumber,
+					}
+				}
 
-	// Loop over our new state changes.
-	// If the record exists in the previous block, add the shares to the existing shares
-	for slotId, newState := range accumulatedState {
-		prepared := &StakerSharesDiff{
-			Staker:      newState.Staker,
-			Strategy:    newState.Strategy,
-			Shares:      newState.Shares,
-			BlockNumber: blockNumber,
-		}
+				var err error
+				currStakerShares, err = stakerShares.ToStakerSharesDiff()
+				if err != nil {
+					return nil, err
+				}
+			}
 
-		if existingRecord, ok := mappedRecords[slotId]; ok {
-			existingShares, success := numbers.NewBig257().SetString(existingRecord.Shares, 10)
+			// add the shares to the existing shares
+			shareDeltaShares, success := new(big.Int).SetString(shareDelta.Shares, 10)
 			if !success {
-				ss.logger.Sugar().Errorw("Failed to convert existing shares to big.Int",
-					zap.String("shares", existingRecord.Shares),
-					zap.String("staker", existingRecord.Staker),
-					zap.String("strategy", existingRecord.Strategy),
-					zap.Uint64("blockNumber", blockNumber),
+				return nil, xerrors.Errorf("Failed to convert shares to big.Int: %s", shareDelta.Shares)
+			}
+			currStakerShares.Shares = currStakerShares.Shares.Add(currStakerShares.Shares, shareDeltaShares)
+
+			// update the local cache
+			updatedStakerShares[slotID] = currStakerShares
+		case *SlashDiff:
+			slashDiff := diff.Event.(*SlashDiff)
+			query := `
```

Comment:
Same with this - put this in a function of its own to make this less big

Reply:
https://github.com/Layr-Labs/sidecar/pull/68/commits/849bd0db77bd553611d6ccdf776de875882dc9c5